homeassistant/components/denonavr/strings.json

```diff
@@ -37,5 +32,18 @@
       "not_denonavr_missing": "Not a Denon AVR Network Receiver, discovery information not complete",
       "no_mac": "Could not retrieve the MAC address from the IP address"
     }
+  },
+  "options": {
+    "step": {
+      "init": {
+        "title": "Denon AVR Network Receivers",
+        "description": "Specify optional settings, a restart is required to apply the settings",
```

Comment:
We have the config entry listener to reload the entry, so a restart isn't needed, right?

Reply:
Yes I will remove that, that was before I used the config entry listener.

One question about this: when I add an extra zone everything works fine and the extra entity is created. However if I delete a zone, the entity of that zone will show as unavailable but still be there. If I then restart HomeAssistant it will show as an orphan entity (or wheter it is called with one of those red exclamation marks "!") and I then need to hit the "remove entity" to get rid of it.

Am I doing something wrong here?
Is there a better way to completly get rid of the entity inside the config entry listener besides just reloading the integration and not setting up the entity anymore?

Reply:
If we change what entities the integration creates we also need to clean up the registries, entity and device registries. We have these methods available:

https://github.com/home-assistant/core/blob/3cc94f7d6a0364e7347bc526d5e9251f8f2b8169/homeassistant/helpers/entity_registry.py#L285

https://github.com/home-assistant/core/blob/3cc94f7d6a0364e7347bc526d5e9251f8f2b8169/homeassistant/helpers/device_registry.py#L322

Reply:
@MartinHjelmare thanks!
The device_registry will not be needed since there is only one device that is always there, only the 2nd and 3th zones entities can be added/removed but they belong to the same device as the main zone entity which is always there.

I will add the async_remove function this evening.

Is there an easy way to check which entity is beeing removed?
possibilities:
 - compare the old and new config_entry.options, but how do I get the old config_entry.options
 - check if the zone 2 and 3 entities exist and compare to the new config_entry.options, but how to properly check if an entitie exists?

(I can probably make it work, but that will probably not be the best way to do it)

Reply:
We can get the existing entity registry entries for a device:
https://github.com/home-assistant/core/blob/1b6f0b78aa7b70a70d16de4b0a22b302b7f4b0d9/homeassistant/helpers/entity_registry.py#L507-L510

Then remove all the entries that doesn't match with the new zones.

Reply:
Nice, thanks!
will do that

Reply:
Done and tested to work, thanks for your help!